### DISCUSSION TITLE
OCSP CDP and peer certificate validation

### DISCUSSION THREAD

**Discussion by amdrsantos** [strongswan/strongswan#1034] [Q&A] [answered]
From what I can see from the documentation OCSP CDP can be configured to use in certificate validation.
https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html#_authorities
Still, I have some questions regarding this functionality:

1. Is this verification being done to X509 certificates received from the peer as well?

2. If the X509 certificate received from the peer, points in its metadata to a CDP and/or OCSP, is strongSwan also getting the CRLs from the servers pointed in the metadata?
Example:
            X509v3 CRL Distribution Points:
                Full Name:
                  URI:http://crl.globalsign.net/root.crl
            Authority Information Access:
                OCSP - URI:http://ocsp.globalsign.com/rootr1

3. Is there a need to use other plugins to connect with the CDP and OCSP servers?

Thanks in Advance,
Alex

**Comment by tobiasbrunner** [comment, marked as answer]
> Is this verification being done to X509 certificates received from the peer as well?

Only to them. Any certificate locally installed is ultimately trusted and not verified via OCSP or CRLs.

> If the X509 certificate received from the peer, points in its metadata to a CDP and/or OCSP, is strongSwan also getting the CRLs from the servers pointed in the metadata?

Yes, but only if the explicitly configured URLs don't yield a result. OCSP servers (from authority sections or certificates) are preferred over CRLs (from authority sections or certificates). If OCSP is successful (i.e. we know that the certificate is valid or revoked), no CRLs are fetched.

> Is there a need to use other plugins to connect with the CDP and OCSP servers?

Yes, you need a "fetcher" plugin (e.g. the [_curl_ plugin](https://docs.strongswan.org/docs/5.9/plugins/curl.html)). The validation itself is done by the _revocation_ plugin, so you need that too (it's enabled, by default, though).

**Reply by amdrsantos** [reply]
Thanks Tobias,

> Yes, but only if the explicitly configured URLs don't yield a result.

What happens if the one OCSP server yield a result, does it continues to the other servers?

Alex

**Reply by tobiasbrunner** [reply]
No, if the OCSP server returns a trusted and valid response that contains a status (good or revoked) for the certificate in question, that's the accepted result. No other servers (or, as mentioned, CRLs) will be checked.

**Reply by amdrsantos** [reply]
> No, if the OCSP server returns a trusted and valid response that contains a status (good or revoked) for the certificate in question, that's the accepted result. No other servers (or, as mentioned, CRLs) will be checked.

My question was more to know what strongSwan does if the server does **not** return a result. Does it proceed to check the next server?

**Reply by tobiasbrunner** [reply]
Yes, it checks all available URLs until it finds a result (see my original reply for the order). Note that no URLs are checked if a cached OCSP response (or CRL) that's still valid is found (revoked states in those are even applied if the cached response/CRL is stale).

**Reply by amdrsantos** [reply]
One more question:
If I disable the CRL validation in revocation.conf `enable_crl = no`, will strongSwan still check the certificate against the CRLs from the CDPs defined in the certificate metadata, or it does not do any CRL validation at all?
Thanks

**Reply by tobiasbrunner** [reply]
Why is this confusing?

**Reply by amdrsantos** [reply]
I just wanted to confirm the same behavior for the CRLs in for the `<name>.crl_uris` field and the URIs in the metadata of peer certificates.

**Reply by tobiasbrunner** [reply]
Why would you assume that's not the case? Or that the option is anything but absolute?

**Reply by amdrsantos** [reply]
No reason. Thanks.
